cloud/logstream.go

```diff
@@ -5,69 +5,190 @@ import (
 	"io"
 	"sync/atomic"
 
-	"github.com/earthly/cloud-api/logstream"
-	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	pb "github.com/earthly/cloud-api/logstream"
+	"github.com/earthly/earthly/util/stringutil"
+	"github.com/hashicorp/go-multierror"
 	"github.com/pkg/errors"
 	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
-// Deltas is a type for iterating over logstream deltas.
-type Deltas interface {
-	Next(ctx context.Context) ([]*logstream.Delta, error)
+func (c *Client) StreamLogs(ctx context.Context, man *pb.RunManifest, ch <-chan *pb.Delta, verbose bool) error {
+	if man.GetResumeToken() == "" {
+		man.ResumeToken = stringutil.RandomAlphanumeric(40)
+	}
+	var (
+		errs  []error
+		retry bool
+		last  *pb.Delta
+	)
+	for {
+		first := []*pb.Delta{firstDelta(man, retry)}
+		if last != nil {
+			first = append(first, last)
+		}
+		var err error
+		last, err = c.streamLogsAttempt(ctx, man.GetBuildId(), first, ch)
+		if err != nil {
+			if retryable(err) {
+				retry = true
+				if verbose {
+					errs = append(errs, err)
+				}
+				continue
```

Comment:
is there any exponential backoff in this retry loop?

Reply:
Not at the moment. It's probably worth adding something simple to prevent spamming.